vervet-underground/internal/storage/mem/mem.go

```diff
@@ -6,81 +6,274 @@ package mem
 // HasVersion implements scraper.Storage.
 func (s *Storage) HasVersion(ctx context.Context, name string, version string, digest string) (bool, error) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	digests, ok := s.contents[serviceVersion{service: name, version: version}]
+	revisions, ok := s.serviceVersionMappedRevisionSpecs[name][version]
+
 	if !ok {
 		return false, nil
 	}
-	_, ok = digests[digest]
+	_, ok = revisions[storage.Digest(digest)]
 	return ok, nil
 }
 
 // NotifyVersion implements scraper.Storage.
 func (s *Storage) NotifyVersion(ctx context.Context, name string, version string, contents []byte, scrapeTime time.Time) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	k := serviceVersion{service: name, version: version}
-	digest := storage.Digest(contents)
-	if digests, ok := s.contents[k]; ok {
-		if _, ok := digests[digest]; ok {
+
+	digest := storage.NewDigest(contents)
+
+	parsedVersion, err := vervet.ParseVersion(version)
+	if err != nil {
+		log.Error().Err(err).Msgf("Failed to resolve Vervet version for %s : %s", name, version)
+		return err
+	}
+
+	// Check if service and version structures are initialized
+	if _, ok := s.serviceVersionMappedRevisionSpecs[name]; !ok {
+		s.serviceVersionMappedRevisionSpecs[name] = versionMappedRevisionSpecs{}
+	}
+
+	revisions, ok := s.serviceVersionMappedRevisionSpecs[name][version]
+	if ok {
+		if _, ok := revisions[digest]; ok {
 			return nil
 		}
 	} else {
-		s.contents[k] = map[string][]byte{}
+		s.serviceVersionMappedRevisionSpecs[name][version] = mappedRevisionSpecs{}
+		revisions = s.serviceVersionMappedRevisionSpecs[name][version]
+	}
+
+	// If the version is newly initialized, meaning no revisions exist,
+	// create the new vervet.VersionSlice with that version initialized
+	// else, append it to the existing service's VersionSlice
+	if len(revisions) == 0 {
+		if _, ok = s.serviceVersions[name]; !ok {
+			s.serviceVersions[name] = vervet.VersionSlice{*parsedVersion}
+		} else {
+			s.serviceVersions[name] = append(s.serviceVersions[name], *parsedVersion)
+			// sort versions when new ones are introduced to maintain BST functionality
+			sort.Sort(s.serviceVersions[name])
+		}
+	}
+	// End of initializations
+
+	// add the new ContentRevision
+	s.serviceVersionMappedRevisionSpecs[name][version][digest] = ContentRevision{
+		serviceVersion: name + "_" + version,
+		timestamp:      scrapeTime,
+		digest:         digest,
+		blob:           contents,
+	}
+
+	return nil
+}
+
+// CollateVersions does the following:
+//   - calls updateCollatedVersions for a slice of unique vervet.Version entries
+//   - for each unique vervet.Version, run collateVersion to create a compiled VU openapi doc
+func (s *Storage) CollateVersions() error {
+	var errs error
+	s.updateCollatedVersions()
+	for _, version := range s.collatedVersions {
+		err := s.collateVersion(version)
+		if err != nil {
+			log.Error().Err(err).Msgf("Could not Collate Version %s", version.String())
+			errs = multierr.Append(errs, err)
```

Comment:
Minor suggestion, but might be more compact and less chatty to 

```suggestion
			errs = multierr.Append(errs, errors.Wrapf(err, "failed to collate version %s", version.String())
```

and let the caller log the error? `multierr` should show all of them when you log it.